File: tp-8/Nave/haskell/Tripulante.hs
module Tripulante (
    Tripulante,
    crearT,
    asignarS,
    sectoresT,
    nombre,
    rango
) where

import Set
import Tipos


data Tripulante = T Nombre Rango (Set SectorId) deriving (Show, Eq, Ord)

{-
    Proposito:
    PRECONDICION:
    COSTO: O(1) es constante
-}
crearT :: Nombre -> Rango -> Tripulante
crearT n r = T n r emptyS

{-
    Proposito:
    PRECONDICION:
    COSTO: O(log S) siendo S la cantidad de Sectores del Set
-}
asignarS :: SectorId -> Tripulante -> Tripulante
asignarS s (T n r ss) = T n r (addS s ss)

{-
    Proposito:
    PRECONDICION:
    COSTO: O(1) es constante
-}
sectoresT :: Tripulante -> Set SectorId
sectoresT (T n r ss) = ss

{-
    Proposito:
    PRECONDICION:
    COSTO: O(1) es constante
-}
nombre :: Tripulante -> Nombre
nombre (T n r ss) = n

{-
    Proposito:
    PRECONDICION:
    COSTO: O(1) es constante
-}
rango :: Tripulante -> Rango
rango (T n r ss) = r

File: tp-8/Nave/cplusplus/tripulante/Tripulante.cpp

#include <iostream>
#include "Tripulante.h"
#include "../set/Set.h"

///costo: constante O(1)
Tripulante crearT( Nombre n, Rango r){
    TripulanteSt* t = new TripulanteSt;
    t->sectores = emptyS();
    t->nombre = n;
    t->rango = r;
    return t;
}

//costo: O(n) donde n es la cantidad de sectores asignados al tripulante
void asignarS(SectorId s, Tripulante t){
    Set sectores = t->sectores;
    addS(s, sectores);
    t->sectores = sectores;
}

//costo: constante O(1)
Set sectoresT(Tripulante t){
    return t->sectores;
}

//costo: constante O(1)
Nombre nombre(Tripulante t){
    return t->nombre;
}

//Coasto: constante O(1)
Rango rango(Tripulante t){
    return t->rango;
}

int testTripulante() {

    std::cout << "=== Ejemplo de uso de Tripulante ===" << std::endl;
    Tripulante tripulante = crearT("Juan", 3);


    std::cout << "Nombre: " << nombre(tripulante) << std::endl;
    std::cout << "Rango: " << rango(tripulante) << std::endl;

    asignarS(1, tripulante);
    asignarS(2, tripulante);
    std::cout << "Sectores asignados size: " << sizeS(sectoresT(tripulante)) << std::endl;

    std::cout << "Sectores asignados: ";
    showSet(sectoresT(tripulante));
    std::cout << "=== Fin del ejemplo ===" << std::endl;
    return 0;
}